config.go

```diff
@@ -3,83 +3,91 @@ package main
+	User         string `envconfig:"USER" default:"haproxy"`
+	Group        string `envconfig:"GROUP" default:"haproxy"`
+	UseHostnames bool   `envconfig:"USE_HOSTNAMES" default:"false"`
 }
 
 type ServicesConfig struct {
-	NameMatch    string `toml:"name_match"`
-	ServiceNamer string `toml:"namer"`
-	NameLabel    string `toml:"name_label"`
+	NameMatch    string `envconfig:"NAME_MATCH"`
+	ServiceNamer string `envconfig:"NAMER" default:"docker_label"`
+	NameLabel    string `envconfig:"NAME_LABEL" default:"ServiceName"`
 }
 
 type SidecarConfig struct {
-	ExcludeIPs           []string `toml:"exclude_ips"`
-	Discovery            []string `toml:"discovery"`
-	StatsAddr            string   `toml:"stats_addr"`
-	PushPullInterval     duration `toml:"push_pull_interval"`
-	GossipMessages       int      `toml:"gossip_messages"`
-	LoggingFormat        string   `toml:"logging_format"`
-	LoggingLevel         string   `toml:"logging_level"`
-	DefaultCheckEndpoint string   `toml:"default_check_endpoint"`
+	ExcludeIPs           []string      `envconfig:"EXCLUDE_IPS" default:"192.168.168.168"`
+	Discovery            []string      `envconfig:"DISCOVERY" default:"docker"`
+	StatsAddr            string        `envconfig:"STATS_ADDR"`
+	PushPullInterval     time.Duration `envconfig:"PUSH_PULL_INTERVAL" default:"20s"`
+	GossipMessages       int           `envconfig:"GOSSIP_MESSAGES" default:"15"`
+	LoggingFormat        string        `envconfig:"LOGGING_FORMAT"`
+	LoggingLevel         string        `envconfig:"LOGGING_LEVEL" default:"info"`
+	DefaultCheckEndpoint string        `envconfig:"DEFAULT_CHECK_ENDPOINT" default:"/version"`
 }
 
 type DockerConfig struct {
-	DockerURL string `toml:"docker_url"`
+	DockerURL string `envconfig:"URL" default:"unix:///var/run/docker.sock"`
 }
 
 type StaticConfig struct {
-	ConfigFile string `toml:"config_file"`
+	ConfigFile string `envconfig:"CONFIG_FILE" default:"static.json"`
 }
 
 type Config struct {
-	Sidecar         SidecarConfig      `toml:"sidecar"`
-	DockerDiscovery DockerConfig       `toml:"docker_discovery"`
-	StaticDiscovery StaticConfig       `toml:"static_discovery"`
-	Services        ServicesConfig     `toml:"services"`
-	HAproxy         HAproxyConfig      `toml:"haproxy"`
-	Listeners       ListenerUrlsConfig `toml:"listeners"`
+	Sidecar         SidecarConfig      // SIDECAR_
+	DockerDiscovery DockerConfig       // DOCKER_
+	StaticDiscovery StaticConfig       // STATIC_
+	Services        ServicesConfig     // SERVICES_
+	HAproxy         HAproxyConfig      // HAPROXY_
+	Listeners       ListenerUrlsConfig // LISTENERS_
 }
 
-func setDefaults(config *Config) {
-	config.DockerDiscovery.DockerURL = "tcp://localhost:2375"
-	config.StaticDiscovery.ConfigFile = "static.json"
-}
+func parseConfig(path string) Config {
+	var config Config
 
-type duration struct {
-	time.Duration
-}
+	err := envconfig.Process("sidecar", &config.Sidecar)
+	if err != nil {
+		exitWithError(err, "Failed to parse environment config")
```

Comment:
"Failed to parse sidecar environment config"

Reply:
I'm not sure if it's at all meaningful to people which section of the struct failed to parse. I had thought of just building up a slice of errors and putting this message at the bottom. Possibly if I did that and also printed a Rubberneck config, people could better figure out where it went wrong.

Reply:
I just pushed a commit with that change.